### DISCUSSION TITLE
Could this restrict searches to test vs non-test code (in Rust)?

### DISCUSSION THREAD

**Discussion by dandavison** [osa1/sg#6] [Q&A]
Hi, this looks like a great project. I was wondering whether any tools exist that can restrict searches in Rust code to test vs non-test code (i.e. according to `#[cfg(test)]`), and came across this project. I haven't looked into tree-sitter yet -- is that something that could be done in this project in principle?

**Comment by osa1** [comment]
Hi @dandavison. Thanks for the question, I think this is an interesting idea. This kind of thing can be done in a "best-effort" way, i.e. it won't be perfect, but may work in majority of the cases.

Any Rust parser can be used to distinguish a function like

```rust
#[test]
fn f() { ... }
```

from

```rust
fn g() { ... }
```

because the `#[test]` attribute will be visible in the AST. However using only the AST for excluding test code will fail to handle cases like:

```rust
#[cfg(test)]
mod my_tests;
```

In this case the sg won't skip `my_tests`, as for that it needs to resolve the module path, and also scan the modules in top-down order. So now you're into static analysis territory. It would be great to have this kind of thing in e.g. rust-analyzer, but it seems out of scope for sg, which aims to work on incomplete ASTs and does not have any semantic information about the supported languages.

I think it may be good idea to add this feature to sg. I will see if I can spare some time to implement this.
